xilem/src/property_tuple.rs

```diff
@@ -0,0 +1,182 @@
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8; P9, 9);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8; P9, 9; P10, 10);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8; P9, 9; P10, 10; P11, 11);
+
+// ---
+
+// We expect to use the ${index} metavariable here once it's stable
+// https://veykril.github.io/tlborm/decl-macros/minutiae/metavar-expr.html
+
+/// Macro used to declare a list of properties for an element.
+///
+/// The expected way to invoke this macro is:
+///
+/// ```ignore
+/// declare_property_tuple!(
+///     MyWidgetProps;
+///     MyWidget;
+///
+///     SomeProp, 0;
+///     OtherProp, 1;
+///     AnotherProp, 2;
+///     YetOtherProp, 3;
+///     // ...
+/// );
+/// ```
+///
+/// This will declare a type `MyWidgetProp` as an alias to `(Option<SomeProp>, Option<OtherProp>, ...)`.
+///
+/// This will also implement [`HasProperty<Prop>`](crate::style::HasProperty) for `MyWidget` with each of the listed properties.
+///
+/// If `MyWidget` is a generic type with type params `T`, `U`, `V`, you should invoke the macro like this:
+///
+/// ```ignore
+/// declare_property_tuple!(
+///     MyWidgetProps;
+///     MyWidget<T, U, V>;
+///
+///     SomeProp, 0;
+///     OtherProp, 1;
+///     AnotherProp, 2;
+///     YetOtherProp, 3;
+///     // ...
+/// );
+/// ```
+#[macro_export]
+macro_rules! declare_property_tuple {
+    (
+        $Props: ident ;
+        $Self:ident $( < $($Args:ident),* > )? ;
+
+        $(
+            $Type: ident, $idx: tt;
+        )+
+    ) => {
+        $crate::__declare_property_tuple_loop!(
+            $Props; ( $Self $( <$($Args),*> )? ); $($Type, $idx;)+
+        );
+    }
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __declare_property_tuple_loop {
+    (
+        $Props: ident ;
+        $Self:tt;
+
+        $(
+            $Type: ident, $idx: tt;
+        )+
+    ) => {
+        type $Props = ($(Option<$Type>,)+);
+
+        $(
+            $crate::__declare_property_tuple_inner!($Self; $Type; $idx;);
+        )+
+
+    }
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __declare_property_tuple_inner {
+    (
+        (
+            $Self:ident $( < $($Args:ident),* > )?
+        );
+        $Type: ident; $idx: tt;
+    )
+    =>
+    {
+        impl $( <$($Args,)*> )? $crate::style::HasProperty<$Type> for $Self $( <$($Args),*> )? {
+            fn property(&mut self) -> &mut Option<$Type> {
+                    &mut self.properties().$idx
+            }
+        }
+    };
+}
```

Comment:
My macro-foo isn't great - I'm not sure why these two macros exist. I get the impression that they both could be inlined?

Reply:
Hmm. I'm not going to claim that I completely follow, but I believe your claim. (I'm struggling to follow what two things you're looping over, I only see the one loop), but I haven't dug into it thoroughly.

In this kind of case, I'd expect them to just be different branches of the same macro. E.g.:
```rust
macro_rules! impl_for_all {
    ([$($ty:ty),*], $second_arg:tt) => {
        $(
            impl_for_all!(@macro_implementation_detail: $ty, $second_arg);
        )*
    }
    (@macro_implementation_detail: $ty:ty, [$($name:ident),*]) => {
            $(
                impl $name<$ty> {}
            )*
    }
}
```

At the very least, I'd expect that to happen for the two completely internal macros. (I can't remember if you can hide entire branches from the docs.)

Reply:
My experience with reading other crates is that the usual pattern is to use an `@` prefix marker for internal branches - e.g. https://docs.rs/bitmask/latest/bitmask/macro.bitmask.html.

But I won't block on it.